File: frontend/lib/core/network/supabase_service.dart
import 'package:logger/logger.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

import '../config/app_config.dart';

/// Global Supabase client instance
final supabase = Supabase.instance.client;

class SupabaseService {
  static final Logger _logger = Logger();

  /// Initialize Supabase
  static Future<void> initialize() async {
    try {
      await Supabase.initialize(
        url: AppConfig.supabaseUrl,
        anonKey: AppConfig.supabaseAnonKey,
        authOptions: const FlutterAuthClientOptions(
          authFlowType: AuthFlowType.pkce,
        ),
        realtimeClientOptions: const RealtimeClientOptions(
          logLevel: RealtimeLogLevel.info,
        ),
        storageOptions: const StorageClientOptions(
          retryAttempts: 10,
        ),
        postgrestOptions: const PostgrestClientOptions(
          schema: 'public',
        ),
      );
      
      _logger.i('Supabase initialized successfully');
    } catch (e) {
      _logger.e('Failed to initialize Supabase: $e');
      rethrow;
    }
  }

  /// Get current user
  static User? getCurrentUser() {
    return supabase.auth.currentUser;
  }

  /// Check if user is authenticated
  static bool get isAuthenticated => getCurrentUser() != null;

  /// Get current session
  static Session? getCurrentSession() {
    return supabase.auth.currentSession;
  }

  /// Listen to auth state changes
  static Stream<AuthState> get authStateChanges {
    return supabase.auth.onAuthStateChange;
  }

  /// Sign in with email and password
  static Future<AuthResponse> signInWithEmail({
    required String email,
    required String password,
  }) async {
    try {
      final response = await supabase.auth.signInWithPassword(
        email: email,
        password: password,
      );
      
      _logger.i('User signed in: ${response.user?.id}');
      return response;
    } catch (e) {
      _logger.e('Sign in failed: $e');
      rethrow;
    }
  }

  /// Sign up with email and password
  static Future<AuthResponse> signUpWithEmail({
    required String email,
    required String password,
    Map<String, dynamic>? userData,
  }) async {
    try {
      final response = await supabase.auth.signUp(
        email: email,
        password: password,
        data: userData,
      );
      
      _logger.i('User signed up: ${response.user?.id}');
      return response;
    } catch (e) {
      _logger.e('Sign up failed: $e');
      rethrow;
    }
  }

  /// Sign in with Google OAuth
  static Future<bool> signInWithGoogle() async {
    try {
      await supabase.auth.signInWithOAuth(
        OAuthProvider.google,
        redirectTo: 'heybills://login-callback',
      );
      
      _logger.i('Initiated Google OAuth sign in');
      return true;
    } catch (e) {
      _logger.e('Google OAuth failed: $e');
      rethrow;
    }
  }

  /// Sign out
  static Future<void> signOut() async {
    try {
      await supabase.auth.signOut();
      _logger.i('User signed out');
    } catch (e) {
      _logger.e('Sign out failed: $e');
      rethrow;
    }
  }

  /// Reset password
  static Future<void> resetPassword({
    required String email,
  }) async {
    try {
      await supabase.auth.resetPasswordForEmail(
        email,
        redirectTo: 'heybills://reset-password',
      );
      
      _logger.i('Password reset email sent to: $email');
    } catch (e) {
      _logger.e('Password reset failed: $e');
      rethrow;
    }
  }

  /// Update user password
  static Future<UserResponse> updatePassword({
    required String newPassword,
  }) async {
    try {
      final response = await supabase.auth.updateUser(
        UserAttributes(password: newPassword),
      );
      
      _logger.i('Password updated for user: ${response.user?.id}');
      return response;
    } catch (e) {
      _logger.e('Password update failed: $e');
      rethrow;
    }
  }

  /// Get user profile from users table
  static Future<Map<String, dynamic>?> getUserProfile(String userId) async {
    try {
      final response = await supabase
          .from('users')
          .select()
          .eq('id', userId)
          .single();

      return response;
    } catch (e) {
      _logger.e('Failed to get user profile: $e');
      rethrow;
    }
  }

  /// Create or update user profile
  static Future<void> upsertUserProfile({
    required String userId,
    required Map<String, dynamic> profileData,
  }) async {
    try {
      await supabase
          .from('users')
          .upsert({
            'id': userId,
            ...profileData,
            'updated_at': DateTime.now().toIso8601String(),
          });

      _logger.i('User profile updated: $userId');
    } catch (e) {
      _logger.e('Failed to update user profile: $e');
      rethrow;
    }
  }

  /// Upload file to storage
  static Future<String> uploadFile({
    required String bucket,
    required String path,
    required dynamic file, // File or Uint8List
    Map<String, String>? metadata,
  }) async {
    try {
      await supabase.storage.from(bucket).upload(
        path,
        file,
        fileOptions: FileOptions(
          upsert: true,
          metadata: metadata,
        ),
      );

      final publicUrl = supabase.storage.from(bucket).getPublicUrl(path);
      
      _logger.i('File uploaded: $path');
      return publicUrl;
    } catch (e) {
      _logger.e('File upload failed: $e');
      rethrow;
    }
  }

  /// Delete file from storage
  static Future<void> deleteFile({
    required String bucket,
    required String path,
  }) async {
    try {
      await supabase.storage.from(bucket).remove([path]);
      _logger.i('File deleted: $path');
    } catch (e) {
      _logger.e('File delete failed: $e');
      rethrow;
    }
  }

  /// Create a realtime subscription
  static RealtimeChannel subscribeToTable({
    required String table,
    String? filter,
    required void Function(PostgresChangePayload) onData,
  }) {
    try {
      final channel = supabase
          .channel('public:$table')
          .onPostgresChanges(
            event: PostgresChangeEvent.all,
            schema: 'public',
            table: table,
            filter: filter != null ? PostgresChangeFilter(
              type: PostgresChangeFilterType.eq,
              column: 'user_id',
              value: filter,
            ) : null,
            callback: onData,
          )
          .subscribe();

      _logger.i('Subscribed to table: $table');
      return channel;
    } catch (e) {
      _logger.e('Subscription failed: $e');
      rethrow;
    }
  }

  /// Remove a realtime subscription
  static Future<void> removeSubscription(RealtimeChannel channel) async {
    try {
      await supabase.removeChannel(channel);
      _logger.i('Subscription removed');
    } catch (e) {
      _logger.e('Failed to remove subscription: $e');
      rethrow;
    }
  }
}

File: backend/src/services/supabaseService.js
/**
 * Supabase Service Layer
 * High-level service functions for Supabase operations
 * Handles authentication, data operations, and storage
 */

const { db } = require('../utils/database');
const { APIError } = require('../../utils/errorHandler');

/**
 * Authentication Service
 */
class AuthService {
  /**
   * Sign up a new user
   * @param {string} email - User email
   * @param {string} password - User password
   * @param {Object} metadata - Additional user metadata
   * @returns {Promise<Object>} Auth result
   */
  async signUp(email, password, metadata = {}) {
    return db.query(
      () => db.client.auth.signUp({
        email,
        password,
        options: {
          data: metadata
        }
      }),
      'user sign up'
    );
  }

  /**
   * Sign in a user
   * @param {string} email - User email
   * @param {string} password - User password
   * @returns {Promise<Object>} Auth result
   */
  async signIn(email, password) {
    return db.query(
      () => db.client.auth.signInWithPassword({ email, password }),
      'user sign in'
    );
  }

  /**
   * Sign in with OAuth provider
   * @param {string} provider - OAuth provider (google, facebook, etc.)
   * @param {Object} options - Additional options
   * @returns {Promise<Object>} Auth result
   */
  async signInWithOAuth(provider, options = {}) {
    return db.query(
      () => db.client.auth.signInWithOAuth({
        provider,
        options: {
          redirectTo: options.redirectTo || process.env.FRONTEND_URL,
          ...options
        }
      }),
      `OAuth sign in with ${provider}`
    );
  }

  /**
   * Sign out current user
   * @returns {Promise<Object>} Sign out result
   */
  async signOut() {
    return db.query(
      () => db.client.auth.signOut(),
      'user sign out'
    );
  }

  /**
   * Get current user session
   * @returns {Promise<Object>} Session data
   */
  async getSession() {
    return db.query(
      () => db.client.auth.getSession(),
      'get user session'
    );
  }

  /**
   * Refresh user session
   * @returns {Promise<Object>} Refreshed session
   */
  async refreshSession() {
    return db.query(
      () => db.client.auth.refreshSession(),
      'refresh user session'
    );
  }

  /**
   * Reset user password
   * @param {string} email - User email
   * @returns {Promise<Object>} Reset result
   */
  async resetPassword(email) {
    return db.query(
      () => db.client.auth.resetPasswordForEmail(email, {
        redirectTo: `${process.env.FRONTEND_URL}/auth/reset-password`
      }),
      'password reset request'
    );
  }

  /**
   * Update user password
   * @param {string} newPassword - New password
   * @returns {Promise<Object>} Update result
   */
  async updatePassword(newPassword) {
    return db.query(
      () => db.client.auth.updateUser({ password: newPassword }),
      'password update'
    );
  }

  /**
   * Verify user from JWT token
   * @param {string} token - JWT token
   * @returns {Promise<Object>} User data
   */
  async verifyToken(token) {
    return db.query(
      () => db.client.auth.getUser(token),
      'token verification'
    );
  }
}

/**
 * User Management Service
 */
class UserService {
  /**
   * Get user profile by ID
   * @param {string} userId - User ID
   * @returns {Promise<Object>} User profile
   */
  async getUserProfile(userId) {
    return db.query(
      () => db.buildQuery('users', {
        select: '*',
        filters: { id: userId }
      }).single(),
      'get user profile'
    );
  }

  /**
   * Update user profile
   * @param {string} userId - User ID
   * @param {Object} updates - Profile updates
   * @returns {Promise<Object>} Updated profile
   */
  async updateUserProfile(userId, updates) {
    const sanitizedUpdates = {
      ...updates,
      updated_at: new Date().toISOString(),
      // Remove sensitive fields that shouldn't be updated directly
      id: undefined,
      created_at: undefined,
      email_confirmed_at: undefined
    };

    return db.query(
      () => db.client
        .from('users')
        .update(sanitizedUpdates)
        .eq('id', userId)
        .select()
        .single(),
      'update user profile'
    );
  }

  /**
   * Create user profile (usually called after auth signup)
   * @param {Object} userData - User data
   * @returns {Promise<Object>} Created profile
   */
  async createUserProfile(userData) {
    const profileData = {
      id: userData.id,
      email: userData.email,
      full_name: userData.user_metadata?.full_name,
      avatar_url: userData.user_metadata?.avatar_url,
      provider: userData.app_metadata?.provider,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString()
    };

    return db.query(
      () => db.client
        .from('users')
        .insert(profileData)
        .select()
        .single(),
      'create user profile'
    );
  }

  /**
   * Delete user profile
   * @param {string} userId - User ID
   * @returns {Promise<Object>} Delete result
   */
  async deleteUserProfile(userId) {
    return db.adminQuery(
      (adminClient) => adminClient
        .from('users')
        .delete()
        .eq('id', userId),
      'delete user profile'
    );
  }

  /**
   * Get user statistics
   * @param {string} userId - User ID
   * @returns {Promise<Object>} User statistics
   */
  async getUserStats(userId) {
    const [receiptsResult, warrantiesResult, budgetsResult] = await Promise.all([
      db.query(
        () => db.client
          .from('receipts')
          .select('count')
          .eq('user_id', userId),
        'get receipts count'
      ),
      db.query(
        () => db.client
          .from('warranties')
          .select('count')
          .eq('user_id', userId),
        'get warranties count'
      ),
      db.query(
        () => db.client
          .from('budgets')
          .select('count')
          .eq('user_id', userId),
        'get budgets count'
      )
    ]);

    return {
      receipts_count: receiptsResult.count || 0,
      warranties_count: warrantiesResult.count || 0,
      budgets_count: budgetsResult.count || 0,
      member_since: (await this.getUserProfile(userId)).data?.created_at
    };
  }
}

/**
 * Storage Service
 */
class StorageService {
  /**
   * Upload file to Supabase Storage
   * @param {string} bucket - Storage bucket name
   * @param {string} path - File path
   * @param {File|Buffer} file - File to upload
   * @param {Object} options - Upload options
   * @returns {Promise<Object>} Upload result
   */
  async uploadFile(bucket, path, file, options = {}) {
    return db.query(
      () => db.client.storage
        .from(bucket)
        .upload(path, file, {
          cacheControl: '3600',
          upsert: false,
          ...options
        }),
      `upload file to ${bucket}/${path}`
    );
  }

  /**
   * Download file from Supabase Storage
   * @param {string} bucket - Storage bucket name
   * @param {string} path - File path
   * @returns {Promise<Object>} Download result
   */
  async downloadFile(bucket, path) {
    return db.query(
      () => db.client.storage
        .from(bucket)
        .download(path),
      `download file from ${bucket}/${path}`
    );
  }

  /**
   * Delete file from Supabase Storage
   * @param {string} bucket - Storage bucket name
   * @param {string} path - File path
   * @returns {Promise<Object>} Delete result
   */
  async deleteFile(bucket, path) {
    return db.query(
      () => db.client.storage
        .from(bucket)
        .remove([path]),
      `delete file from ${bucket}/${path}`
    );
  }

  /**
   * Get public URL for file
   * @param {string} bucket - Storage bucket name
   * @param {string} path - File path
   * @returns {Object} Public URL data
   */
  getPublicUrl(bucket, path) {
    return db.client.storage
      .from(bucket)
      .getPublicUrl(path);
  }

  /**
   * Create signed URL for private file
   * @param {string} bucket - Storage bucket name
   * @param {string} path - File path
   * @param {number} expiresIn - Expiry time in seconds
   * @returns {Promise<Object>} Signed URL
   */
  async createSignedUrl(bucket, path, expiresIn = 3600) {
    return db.query(
      () => db.client.storage
        .from(bucket)
        .createSignedUrl(path, expiresIn),
      `create signed URL for ${bucket}/${path}`
    );
  }

  /**
   * List files in bucket
   * @param {string} bucket - Storage bucket name
   * @param {string} folder - Folder path (optional)
   * @returns {Promise<Object>} File list
   */
  async listFiles(bucket, folder = '') {
    return db.query(
      () => db.client.storage
        .from(bucket)
        .list(folder),
      `list files in ${bucket}/${folder}`
    );
  }
}

/**
 * Real-time Service
 */
class RealtimeService {
  /**
   * Subscribe to table changes
   * @param {string} table - Table name
   * @param {Function} callback - Callback function
   * @param {Object} options - Subscription options
   * @returns {Object} Subscription
   */
  subscribe(table, callback, options = {}) {
    let subscription = db.client
      .channel(`${table}_changes`)
      .on('postgres_changes', {
        event: '*',
        schema: 'public',
        table,
        ...options
      }, callback);

    subscription.subscribe();
    return subscription;
  }

  /**
   * Unsubscribe from changes
   * @param {Object} subscription - Subscription to remove
   */
  unsubscribe(subscription) {
    if (subscription) {
      subscription.unsubscribe();
    }
  }

  /**
   * Send real-time message to channel
   * @param {string} channel - Channel name
   * @param {string} event - Event name
   * @param {Object} payload - Message payload
   */
  async sendMessage(channel, event, payload) {
    const channelInstance = db.client.channel(channel);
    return channelInstance.send({
      type: 'broadcast',
      event,
      payload
    });
  }
}

// Create service instances
const authService = new AuthService();
const userService = new UserService();
const storageService = new StorageService();
const realtimeService = new RealtimeService();

module.exports = {
  AuthService,
  UserService,
  StorageService,
  RealtimeService,
  authService,
  userService,
  storageService,
  realtimeService
};